### PR TITLE
Display R package help in output window rather than in browser

### PR DESCRIPTION
**Pull request by N-thony**
@rdstern this is ready for you to test. I made the changes in the FromLibrary, HelpVignettes, HypothesisTestsCalculator, Modelling, UseModel, and Calculator Dialogues. Some dialogue `Help button` is still `Help` the changes will be make in another PR.
@lloyddewit could you have a look of the code?
Thanks

### CODE REVIEW

**Review by rdstern — COMMENTED** [review]
@N-thony this is excellent.  The only one that doesn't work is in the R Documentation the Vignettes option works in 0.7.8 see below and doesn't do anything in the new version.

image

There is a simple addition needed to the dialogues, on making it R-help, but that's a separate issue.  I'll write it.

**Comment by N-thony** [comment]
> @N-thony this is excellent. The only one that doesn't work is in the R Documentation the Vignettes option works in 0.7.8 see below and doesn't do anything in the new version.
> 
> image
> 
> There is a simple addition needed to the dialogues, on making it R-help, but that's a separate issue. I'll write it.

@rdstern @lloyddewit I found that for the vignette, it autogenerates the html file. I need to investigate more about and it might delay this PR, if you happy we can merge this and fix the remaining in another PR?

**Review by rdstern — APPROVED** [review]
@lloyddewit then let's merge as is, and give many points to anyone who notices this vignette option is missing - even in the next month!.

**Review by rdstern — APPROVED** [review]

**Comment by N-thony** [comment]
> > @lloyddewit then let's merge as is, and give many points to anyone who notices this vignette option is missing - even in the next month!.
> 
> Currently the vignette is still going to the browser,

@lloyddewit I think this can wait a bit, I found a way to display the vignette in the maximize window

**Comment by N-thony** [comment]
> > > @lloyddewit then let's merge as is, and give many points to anyone who notices this vignette option is missing - even in the next month!.
> > 
> > 
> > Currently the vignette is still going to the browser,
> 
> @lloyddewit I think this can wait a bit, I found a way to display the vignette in the maximize window

@rdstern I have added the display of the vignette in the maximize window. Have a look.

**Review by rdstern — DISMISSED** [review]
Great.  Good if this can be included in the February release.

File: instat/UserControls/Webview/Windows/ucrWebViewer.vb [hunk 1]
```diff
@@ -45,8 +45,10 @@ Public Class ucrWebViewer
         'it's not yet clear how we can implement a custom schema at this point,
         'not unless we specify R-Instat temp output folder in the R commands.
         'that should be the first step
-
-        Dim strUrl As String = "file:///" + strFileName.Replace("\", "/")
+        Dim strUrl As String = strFileName
+        If bReplace Then
+            bReplace = "file:///" + strFileName.Replace("\", "/")
```
Comment (lloyddewit) [inline instat/UserControls/Webview/Windows/ucrWebViewer.vb:50]: we are assigning a string to a boolean, is that correct?
should we assign to `strUrl` instead?

File: instat/dlgFromLibrary.vb [hunk 2]
```diff
@@ -229,13 +230,29 @@ Public Class dlgFromLibrary
...
+        clsGetPortFunction.SetPackageName("tools")
+        clsGetPortFunction.SetRCommand("startDynamicHelp")
+        clsGetPortFunction.AddParameter("start", "NA", iPosition:=0)
+
+        Dim expPortTemp As SymbolicExpression
+        expPortTemp = frmMain.clsRLink.RunInternalScriptGetValue(clsGetPortFunction.ToScript(), bSeparateThread:=False)
+        Dim strPort As String = ""
+        If expPortTemp IsNot Nothing AndAlso expPortTemp.Type <> Internals.SymbolicExpressionType.Null Then
+            strPort = expPortTemp.AsInteger(0)
+        End If
```
Comment (lloyddewit) [inline instat/dlgFromLibrary.vb:244]: This code is duplicated 4-5 times below.
Move this code into a public function that can be called by the different dialog boxes?

File: instat/dlgFromLibrary.vb [hunk 3]
```diff
@@ -229,13 +230,29 @@ Public Class dlgFromLibrary
...
+
+        Dim strPackageName As String = ucrInputPackages.cboInput.SelectedItem
+        Dim strTopic As String = lstCollection.SelectedItems(0).Text
+        Dim strFilePath As String = Path.Combine("library", strPackageName, "html", String.Concat(strTopic, ".html"))
+        Dim strLocalHost As String = "127.0.0.1:"
+        Dim strURL As String
+        strURL = Path.Combine(String.Concat("http://", strLocalHost), strPort, strFilePath)
```
Comment (lloyddewit) [inline instat/dlgFromLibrary.vb:251]: combine into single line?

File: instat/dlgHelpVignettes.vb [hunk 4]
```diff
@@ -16,12 +16,15 @@
...
 Imports instat.Translations
+Imports System.IO
 
 Public Class dlgHelpVignettes
     Private bFirstload As Boolean = True
     Private bReset As Boolean = True
     Private strAvailablePackages() As String
-    Private clsHelpFunction, clsVignettesFunction As New RFunction
+    Private clsGetVignetteFunction As New RFunction
+    Private clsGetPortFunction As New RFunction
```
Comment (lloyddewit) [inline instat/dlgHelpVignettes.vb:26]: should this be a class member? make this a local variable in `OpenHelpFile()`?

File: instat/dlgFromLibrary.vb [hunk 5]
```diff
@@ -229,13 +230,29 @@ Public Class dlgFromLibrary
...
+        clsGetPortFunction.AddParameter("start", "NA", iPosition:=0)
+
+        Dim expPortTemp As SymbolicExpression
+        expPortTemp = frmMain.clsRLink.RunInternalScriptGetValue(clsGetPortFunction.ToScript(), bSeparateThread:=False)
+        Dim strPort As String = ""
+        If expPortTemp IsNot Nothing AndAlso expPortTemp.Type <> Internals.SymbolicExpressionType.Null Then
+            strPort = expPortTemp.AsInteger(0)
+        End If
+
+        Dim strPackageName As String = ucrInputPackages.cboInput.SelectedItem
```
Comment (lloyddewit) [inline instat/dlgFromLibrary.vb:246]: the remaining code in this sub, also seems to be duplicated 4-5 times in the dialogs below.
Is it possible to reduce some of this duplication?

**Review by lloyddewit — CHANGES_REQUESTED** [review]

File: instat/dlgHelpVignettes.vb (hunk 4 above)
Comment (N-thony) [reply instat/dlgHelpVignettes.vb]: This seems fixed now

File: instat/dlgFromLibrary.vb (hunk 3 above)
Comment (N-thony) [reply instat/dlgFromLibrary.vb]: This seems fixed now

File: instat/dlgFromLibrary.vb (hunk 2 above)
Comment (N-thony) [reply instat/dlgFromLibrary.vb:244]: This seems fixed now

File: instat/UserControls/Webview/Windows/ucrWebViewer.vb (hunk 1 above)
Comment (N-thony) [reply instat/UserControls/Webview/Windows/ucrWebViewer.vb]: This seems fixed now

File: instat/dlgFromLibrary.vb [hunk 6]
```diff
@@ -228,14 +229,48 @@ Public Class dlgFromLibrary
         End If
     End Function
 
+    Public Function GetFileURL(strPackageName As String, Optional strTopic As String = "", Optional bVignette As Boolean = False) As String
```
Comment (lloyddewit) [inline instat/dlgFromLibrary.vb:232]: Great that you created this function.
Dialogs should be independent from each other. Putting the function in `dlgFromLibrary` means that other dialogs become dependent on `dlgFromLibrary`. 
Could we put the function somewhere else? I'm not aware of any utility class in R-Instat where we could put this type of function, can you think of any?
If not, then we could make a new class (e.g. `clsFileUrlUtilities.vb`) and use the `shared` key word for the function definition.
What's your view?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@N-thony thank you, this looks good. Just one comment. If this can be resolved, then we can merge.

**Comment by lloyddewit** [comment]
@lilyclements Please could you review the R code?

@ChrisMarsh82 I think that @rdstern would like this included in the Feb release. If @lilyclements is not able to review before you want to make the build, and I've already approved this PR, then please go ahead and merge this PR, thanks

(please note, I'm unavailable on Tuesday 28 Feb)

File: instat/dlgFromLibrary.vb (hunk 6 above)
Comment (N-thony) [reply instat/dlgFromLibrary.vb]: @lloyddewit this sounds good to me, I have the function to `clsFileUrlUtilities.vb`

**Review by lloyddewit — APPROVED** [review]

**Review by rdstern — APPROVED** [review]
looks fine now

**Pull request merged**